src/Compilers/Core/Portable/InternalUtilities/SpecializedCollections.cs

```diff
@@ -1,55 +1,32 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 
 namespace Roslyn.Utilities
 {
     internal static partial class SpecializedCollections
     {
-        public static IEnumerator<T> EmptyEnumerator<T>()
-        {
-            return Empty.Enumerator<T>.Instance;
-        }
+        public static IEnumerator<T> EmptyEnumerator<T>() => EmptyEnumerable<T>().GetEnumerator();
 
-        public static IEnumerable<T> EmptyEnumerable<T>()
-        {
-            return Empty.List<T>.Instance;
-        }
+        public static IEnumerable<T> EmptyEnumerable<T>() => Array.Empty<T>();
```

Comment:
❗️ This produces `T[]` as an `IEnumerable<T>`, which does not cache its instance of `IEnumerator<T>`. This will force an allocation on all attempts to enumerate the collection.

Reply:
@sharwell This optimization is present for as long as the code has been on Github. https://github.com/dotnet/coreclr/blob/ef1e2ab328087c61a6878c1e84f4fc5d710aebce/src/mscorlib/src/System/Array.cs#L2717

I can manually check 4.6.1, though.

Reply:
Thank you for checking. I drew the same conclusion as you stated here but wasn't sure how to check a specific desktop framework release.

Reply:
@sharwell I found the reference source for the framework. It looks like this code was added for .NET 4.6. https://github.com/Microsoft/referencesource/blob/ec178a5e7deb87a9cc7e0982ee32b7d965735b16/mscorlib/system/array.cs#L2750

I think that's a good enough guarantee.